File: montecarlo_pi.py
import random
import math


def is_inside_circle():
    x = random.random()
    y = random.random()
    if x**2 + y**2 <= 1:
        return True
    else:
        return False

def calculate_pi(iterations, out):
    inside = 0
    for i in range(iterations):
        if is_inside_circle():
            inside += 1
    pi = 4 * inside / iterations
    
    if not out:
        return pi
    else:
        print(f"{iterations} iterations: {pi} ({'{0:+g}'.format(pi - math.pi)})")


for i in range(2, 10):
    calculate_pi(10**i, True)

File: montecarlo_pi.cpp
#include <iostream>
#include <cstdlib>
#include <cmath>
#include <ctime>

int is_in_circle() {
    double x = ((double) rand()) / (double) RAND_MAX;
    double y = ((double) rand()) / (double) RAND_MAX;

    if (pow(x, 2) + pow(y, 2) < 1) {
        return 1;
    } else {
        return 0;
    }
}

void calculate_pi(unsigned long iterations) {
    unsigned long inside = 0;
    for (int i = 0; i < iterations; i++) {
        if (is_in_circle()) {
            inside++;
        }
    }
    double est_pi = 4 * ((double) inside / iterations);
    std::cout << iterations << " iterations: " << est_pi << " (" << est_pi - M_PI << ")" << std::endl;
}

int main() {
    srand(time(NULL));
    for (int i = 2; i < 9; i++) {
        calculate_pi(pow(10, i));
    }
}
